### DISCUSSION TITLE
Help needed for clock example going backward, and garbage in background

### DISCUSSION THREAD

**Discussion by N0ury** [Bodmer/TFT_eSPI#2927] [Q&A - General] [answered]
Hi, this is not an issue, it's more a newbie question.
I'm using a X-NUCLEO-GFX01M2 STM display board along with a STM NUCLEO-F411RE board.
Display is a ST7789V (2.2" SPI QVGA TFT LCD 320x240).
It has also a 64-Mbit SPI NOR Flash memory for storing graphic images, texts and texture (I don't use it for the moment).
I use these boards under platformIO.
I try to use the clock example.
Clock is displayed, but seconds needle goes backward (ie from right to left).
Furthermore, the background is not clean on the right side.
I attach a screenshot.
![PXL_20231026_140313108 MP](https://github.com/Bodmer/TFT_eSPI/assets/4224717/bf9e6aa9-d5b5-4c29-8a4f-a4423f4ca43c)

The clock is not centered, not sure if this is a problem.
It's my very first test of this lib, and I'm very happy to see the clock.
But I would like to go further and deeper in testing and using the lib.
I would also like why this clock wants me to become younger...
And also why the background is not clean.
Here's my plaformio.ini file, it's very simple:
```
[env]
platform = ststm32
board = nucleo_f411re
framework = arduino
monitor_port = hwgrep://0483:374b
upload_port = hwgrep://0483:374b
monitor_speed = 115200
lib_deps = TFT_eSPI
	SPI
build_flags =
  -D STM32=1
  -D NUCLEO_64_TFT=1
  -D STM_PORTA_DATA_BUS=1
  -D STM_PORTB_DATA_BUS=1
  -D ST7789_DRIVER=1
  -D USER_SETUP_LOADED=1
  -D ILI9163_DRIVER=1                           ; Select ILI9163 driver
  -D TFT_WIDTH=320                              ; Set TFT size
  -D TFT_HEIGHT=240
  -D TFT_MISO=PA6                               ; Define SPI pins
  -D TFT_MOSI=PA7
  -D TFT_SCLK=PA5
  -D TFT_CS=PA9
  -D TFT_DC=PB10                                ; Data/Comand pin
  -D TFT_RST=PA1                                ; Reset pin
  -D SPI_FREQUENCY=27000000                     ; Set SPI frequency

[env:main]
    build_src_filter = +<main.cpp>

[env:ex]
    build_src_filter = +<ex.cpp>
```

I'm not sure what to put as SPI_FREQUENCY, I leaved it unchanged. Is this the problem?

This lib is very promising for me. I really want to check it more.

Thanks for any help.

**Comment by N0ury** [comment]
Reading back my post I realized there was an error, there was a useless line `-D ILI9163_DRIVER=1 `
I have removed it.
The result is not better.
Here it is in this screenshot:
![PXL_20231026_143429919 MP](https://github.com/Bodmer/TFT_eSPI/assets/4224717/2d374ec8-461d-4714-9970-8f762a9d6d7c)
Seconds needle has disappeared, and garbage is still here. Not at the same place.
Colors have changed too.

**Comment by N0ury** [comment]
Sorry for all this noise.
Looking very closely, now seconds needle goes the right way. It's only in very light yellow on a white background, so near invisible.
But garbage is still here.
Half of the problem is solved...

**Comment by N0ury** [comment]
doc says: // For ST7789, ST7735, ILI9163 and GC9A01 ONLY, define the pixel width and height in portrait orientation
so changed to
  -D TFT_WIDTH=240                              ; Set TFT size
  -D TFT_HEIGHT=320
Garbage has gone away...

Clock is not centered vertically and colors are bad.
Is it wanted or is it an issue. I can't say.

**Comment by Bodmer** [comment, marked as answer]
By default the sketch centres the clock in portrait mode, not landscape. The setRotation function defines the screen rotation.

The colours are inverted, (black = white).  Run the included example diagnostic sketch [Colour_Test](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Test%20and%20diagnostics/Colour_Test/Colour_Test.ino), see header and use the sketch to get black and white correct.

You will then have to use ***one*** of these lines in your sketch after tft.init():

tft.invertDisplay( true );
tft.invertDisplay( false );

Your display also has red and green swapped, so include ***one*** of these lines in your setup, compile, upload and try the Colour_Test example again to find the correct option:

-D TFT_RGB_ORDER TFT_RGB=1 ; Colour order Red-Green-Blue
-D TFT_RGB_ORDER TFT_BGR=1 ; Colour order Blue-Green-Red

These problems occur because different suppliers wire the display pixels in different ways.

**Comment by N0ury** [comment]
@Bodmer tft.invertDisplay( false ); solved the colour issue.

-D TFT_RGB_ORDER TFT_RGB=1 ; Colour order Red-Green-Blue
-D TFT_RGB_ORDER TFT_BGR=1 ; Colour order Blue-Green-Red
Both gives an error.
I tried
-D TFT_RGB_ORDER=TFT_RGB
-D TFT_RGB_ORDER=TFT_BGR
None of them changed anything.

It's a little bit late today, but I'll try tomorrow most of the examples. Including Colour_test sketch.
In the meanwhile thanks a lot. I was trying to use TouchGFX from ST but there tools are very heavy.
Furthermore they work well with the simulator, but it's an other story with the hardware.
I prefer to work in CLI mode with VI as the best editor/IDE.

**Comment by N0ury** [comment]
I couldn't wait and despite the need for sleep using Colour sketch, first part of the loop using `tft.invertDisplay( false );` works fine for
```
  tft.fillScreen(TFT_BLACK);
  tft.drawRect(0, 0, tft.width(), tft.height(), TFT_GREEN);
```
But `tft.setTextColor` and `tft.println` does nothing.
It's as if these instructions do not exist.
Rectangle remain unchanged.

What can be the cause?

I've tried hanging colors. It works only for the rectangles and their border.

**Comment by N0ury** [comment]
I didn't read User_Setup.h until the end.
I'm taking time to read docs and examples.
Font 4 works like a charm!
One thing, STM32 pins are displayed with strange names. Negative values sometimes.
But this is completely out of the scope of the lib.

**Comment by N0ury** [comment]
@Bodmer One problem remains.
I use `tft.invertDisplay( false );` and colors are good now.
But the screen is displayed as if it was a 240x240. I use a 320x240 display.
The down part, or right one, depending on the orientation is made of garbage, or previous screen.
I've tried this with a nucleo-f411re and a nucleo-g071rb (from STM), it's the same.
I have this issue with many of the examples I have tried.
What I'm I doing wrong?
Here's an extract from my platformio.ini file
```
build_flags = 
        -D USER_SETUP_LOADED=1
        -D STM32=1
        -D NUCLEO_64_TFT=1
        -D STM_PORTA_DATA_BUS=1
        -D STM_PORTB_DATA_BUS=1
        -D ST7789_DRIVER=1
        -D TFT_WIDTH=240
        -D TFT_HEIGHT=320
        -D TFT_MISO=PA6
        -D TFT_MOSI=PA7
        -D TFT_SCLK=PA5
        -D TFT_CS=PA9
        -D TFT_DC=PB10
        -D TFT_RST=PA1
        -D SPI_FREQUENCY=80000000
        -D TFT_RGB_ORDER=TFT_BGR ; Colour order Red-Green-Blue
        -D LOAD_GLCD=1   ; Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
        -D LOAD_FONT2=1  ; Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
        -D LOAD_FONT4=1  ; Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
        -D LOAD_FONT6=1  ; Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
        -D LOAD_FONT7=1  ; Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
        -D LOAD_FONT8=1  ; Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#       -D LOAD_FONT8N=1 ; Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
        -D LOAD_GFXFF=1  ; FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts
        -D SMOOTH_FONT=1

```
Furthermore, sometimes tft.begin is used and sometimes tft.init. What's the difference?
Thanks in advance for any help.
